wumpus_world.py

```diff
@@ -118,21 +117,154 @@ def _transform_kb(self):
+                literals = []
+                for part in clause[1:]:
+                    literals.extend(get_literals(part))
+                return literals
+            
+            return [clause]
+
+        def remove_literal(clause, literal):
+            """
+            Helper function to remove a specific literal from a clause.
+            """
+            if isinstance(clause, tuple) and clause[0] == 'OR':
+                # Takes all of the clauses in the tuple that aren't the literal to be removed and puts them in a new tuple
+                remaining_literals = tuple(lit for lit in clause[1:] if lit != literal)
+                return ('OR', *remaining_literals) if remaining_literals else None
+            
+            return None if clause == literal else clause
+
+        literals_1 = get_literals(clause_1)
+        literals_2 = get_literals(clause_2)
+
+        resolved_clauses = []
+
+        for literal_1 in literals_1:
+            for literal_2 in literals_2:
+                # If there are complementary literals they must be removed
+                if is_complementary(literal_1, literal_2):
+                    new_clause_1 = remove_literal(clause_1, literal_1)
+                    new_clause_2 = remove_literal(clause_2, literal_2)
+
+                    # If both clauses still contain literals after removal we must combine the remaining literals with 'OR'
+                    if new_clause_1 and new_clause_2:
+                        resolved = ('OR', get_literals(new_clause_1), get_literals(new_clause_2))
+
+                    # If only clause_1 has remaining literals, use it as the resolved clause
+                    elif new_clause_1:
+                        resolved = new_clause_1
+
+                    # If only clause_2 has remaining literals, use it as the resolved clause
+                    elif new_clause_2:
+                        resolved = new_clause_2
+
+                    # If both clauses are empty after removing complementary literals, return an empty clause
+                    else:
+                        resolved = None
+
+                    # If the resolved clause has only one literal, we must remove the 'OR' and keep the single literal
+                    if isinstance(resolved, tuple) and resolved[0] == 'OR' and len(resolved) == 2:
+                        resolved = resolved[1] 
+
+                    # Add the resolved clause to the list of resolved clauses, if it exists
+                    if resolved:
+                        resolved_clauses.append(resolved)
+        
+        return resolved_clauses
+        
+    def sentence_to_cnf(self, sentence):
+        """
+        Converts a propositional logic sentence into CNF using a recursive approach.
+        """
+        # If the sentence is a literal or does not have 'IMPLIES' or 'IFF' in it, the sentence is already in CNF
+        if isinstance(sentence, str) or (sentence[0] not in {'IMPLIES', 'IFF'}):
```

Comment:
I would clean up this a bit and check only for str. You can distinguish between three cases: 

- is it a string? if so, just return it; 
- is it a tuple? apply transformation
- are there tuples inside the tuple? call sentence_to_cnf recursively on the inside tuples.